3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,248 @@
+          entry => {
+            const creativeMessageRouter = entry[1];
+            creativeMessageRouter.sendQueuedMessagesToListeners();
+          });
+      });
+    this.sendReadyMessageToCreative_();
+  }
+
+  /**
+   * Gets a handle to the window (or testing object that mocks window)
+   * @returns {!Window}
+   */
+  getWindow() {
+    return this.win_;
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @private
+   */
+  sendReadyMessageToCreative_() {
+    this.iframeMessagingClient_.sendMessage(MessageTypes.ampAnalytics3pReady,
+      {senderId: this.sentinel_});
+  }
+
+  /**
+   * Sends a message back to the host window
+   * @param {string} type The type of message to send.
+   * @param {Object=} opt_payload The payload of message to send.
+   */
+  sendMessage(type, opt_payload) {
+    this.iframeMessagingClient_.sendMessage(type, opt_payload);
+  }
+}
+
+/**
+ * @private @const {!AmpAnalytics3pMessageRouter}
+ */
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} sentinel
+   * @param {!string} creativeId
+   */
+  constructor(parent, sentinel, creativeId) {
+    /** @private {!Window} */
+    this.parent_ = parent;
+
+    /**
+     * @private {string}
+     */
+    this.sentinel_ = sentinel;
+
+    /**
+     * @private {string}
+     */
+    this.creativeId_ = creativeId;
+
+    /**
+     * @private {string}
+     */
+    this.extraData_ = null;
+
+    /**
+     * @private {!Array<Object<string,*>>}
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * @private {!Array<!function(Array)>}
+     */
+    this.eventListeners_ = [];
+
+    if (this.parent_.getWindow().onNewAmpAnalyticsInstance) {
+      this.parent_.getWindow().onNewAmpAnalyticsInstance(this);
+    } else {
+      dev().error(TAG_, 'Vendor page must implement onNewAmpAnalyticsInstance' +
+        ' prior to loading library script.');
+    }
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!function(!Array)} listener A function that takes an array of event
+   * strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventListeners_.push(listener);
+  }
+
+  /**
+   * Receives a message from a creative for the cross-domain iframe
+   * @param Object<string,*> message The message that was received
+   */
+  receiveMessage(message) {
+    if (message[MessageTypes.ampAnalytics3pNewCreative]) {
+      this.extraData_ = message[MessageTypes.ampAnalytics3pNewCreative];
+    } else {
+      this.receivedMessagesQueue_.push(message);
+    }
+  }
+
+  /**
+   * Sends enqueued messages to the third-party vendor's metrics-collection page
```

Comment:
Done.
A large part of the comment is going to become moot, since we'll now only support one listener.
Changed "sends" to "passes" to eliminate any perceived implication of postMessage().